### PR TITLE
Task residency support

### PR DESCRIPTION
**Pull request by ondrej-smola**
Described partially in [Persistent Volumes](https://mesosphere.github.io/marathon/docs/persistent-volumes.html)
[RAML type description](https://github.com/mesosphere/marathon/blob/48bfffdbb2920732ba726ce19ade6deb00121794/docs/docs/rest-api/public/api/v2/types/app.raml#L23l)

### CODE REVIEW

File: application.go [hunk 1]
```diff
@@ -571,6 +572,13 @@ func (r *Application) EmptyUnreachableStrategy() *Application {
 	return r
 }
 
+func (r *Application) SetResidency(whenLost TaskLostBehaviorType) *Application {
```
Comment (gambol99) [inline application.go]: // SetResidency .. description ..

File: residency.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package marathon
+
+type TaskLostBehaviorType string
```
Comment (gambol99) [inline residency.go]: could we add comment to the export var's

File: residency.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+const (
+	TaskLostBehaviorTypeWaitForever          TaskLostBehaviorType = "WAIT_FOREVER"
+	TaskLostBehaviorTypeRelaunchAfterTimeout TaskLostBehaviorType = "RELAUNCH_AFTER_TIMEOUT"
+)
+
+type Residency struct {
```
Comment (gambol99) [inline residency.go]: same as above ...

File: residency_test.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
+package marathon
```
Comment (gambol99) [inline residency_test.go]: could be add the licence

File: residency.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package marathon
```
Comment (gambol99) [inline residency.go]: needs the default licence text

File: residency.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+	TaskLostBehavior                 TaskLostBehaviorType `json:"taskLostBehavior,omitempty"`
+	RelaunchEscalationTimeoutSeconds int                  `json:"relaunchEscalationTimeoutSeconds,omitempty"`
+}
+
+func (r *Residency) SetTaskLostBehavior(behavior TaskLostBehaviorType) *Residency {
+	r.TaskLostBehavior = behavior
+	return r
+}
+
+func (r *Residency) SetRelaunchEscalationTimeoutSeconds(seconds int) *Residency {
```
Comment (timoreimann) [inline residency.go]: Let's have the method take a `time.Duration`.

File: residency.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+package marathon
+
+import "time"
+
+// TaskLostBehaviorType sets action taken when the resident task is lost
+type TaskLostBehaviorType string
+
+const (
+	// TaskLostBehaviorTypeWaitForever sets to don`t take any action when the resident task is lost
```
Comment (timoreimann) [inline residency.go]: I'd rephrase this to:

`TaskLostBehaviorTypeWaitForever indicates to not take any action when the resident task is lost`

File: residency.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+	TaskLostBehaviorTypeWaitForever TaskLostBehaviorType = "WAIT_FOREVER"
+	// TaskLostBehaviorTypeRelaunchAfterTimeout sets to try relaunch the lost resident task
+	// on another node after relaunch escalation timeout
```
Comment (timoreimann) [inline residency.go]: Rephrase proposal:

`TaskLostBehaviorTypeWaitForever indicates to try relaunching the lost resident task on another node after the relaunch escalation timeout has elapsed`

File: residency_test.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package marathon
+
+import (
+	"testing"
+
```
Comment (timoreimann) [inline residency_test.go]: No blank line.

File: application.go [hunk 10]
```diff
@@ -99,6 +99,7 @@ type Application struct {
 	LastTaskFailure       *LastTaskFailure        `json:"lastTaskFailure,omitempty"`
 	Fetch                 *[]Fetch                `json:"fetch,omitempty"`
 	IPAddressPerTask      *IPAddressPerTask       `json:"ipAddress,omitempty"`
+	Residency             *Residency              `json:"residence,omitempty"`
```
Comment (timoreimann) [inline application.go]: Should the JSON name really be `residence` or `residency` rather?

Comment (ondrej-smola) [inline application.go]: Ah, thanks, i just switched to another task before be able to test it. Will fix it and do some testing to make sure it works as expected.

**Comment by timoreimann** [comment]
Apart from a missing integration test with our YAML spec, this looks good to me.

**Comment by ondrej-smola** [comment]
I would expect during integration test that if i create some app (fake-app) with one definition, that definition would be returned on all tested endpoints - as marathon would.

**Comment by timoreimann** [comment]
I see the integration test definitions as tailored for one or more particular tests. In my world, they don't have to be equal everywhere.

If you feel strong about it, then I don't mind.

**Comment by ondrej-smola** [comment]
We can go both ways - but we should state this somewhere in contribution guidelines to make it clear for future contributions.

**Comment by ondrej-smola** [comment]
any other issues before merge?

**Comment by timoreimann** [comment]
@gambol99 your turn. :-)

**Comment by ondrej-smola** [comment]
@gambol99 ping

**Comment by gambol99** [comment]
needs a rebase ... but lgtm!!

File: residency.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2017 Ondrej Smola All rights reserved.
```
Comment (timoreimann) [inline residency.go]: is it okay for the copyright to carry the contributor's name? We tend to use @gambol99's even for contributed files.

Really genuine question -- I'm far from being skilled on the licensing level.

Comment (ondrej-smola) [reply residency.go]: I dont care ... will update it to gambol99

Comment (gambol99) [reply residency.go]: I think we should just change this to "Go-Marathon Authors" or alike now? ... we can raise a PR which changes them all

Comment (timoreimann) [reply residency.go]: SGTM! I'll take care of it.

Comment (timoreimann) [reply residency.go]: I took the liberty of changing the name myself here in order to simplify a project-wide rename later, as discussed.

Hope that's okay.
